src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,525 @@
+            RangePayment calldata rangePayment = rangePayments[i];
+            uint256 nonce = paymentNonce[msg.sender];
+            bytes32 rangePaymentForAllHash = keccak256(abi.encode(msg.sender, nonce, rangePayment));
+            require(
+                !isRangePaymentForAllHash[msg.sender][rangePaymentForAllHash],
+                "PaymentCoordinator._payForRange: range payment hash already submitted"
+            );
+            isRangePaymentForAllHash[msg.sender][rangePaymentForAllHash] = true;
+
+            _payForRange(rangePayment);
+            emit RangePaymentForAllCreated(msg.sender, nonce, rangePaymentForAllHash, rangePayment);
+            paymentNonce[msg.sender] = nonce + 1;
+        }
+    }
+
+    /**
+     * @notice Claim payments against a given root (read from distributionRoots[claim.rootIndex]).
+     * Earnings are cumulative so earners don't have to claim against all distribution roots they have earnings for,
+     * they can simply claim against the latest root and the contract will calculate the difference between
+     * their cumulativeEarnings and cumulativeClaimed. This difference is then transferred to claimerFor[claim.earner]
+     * @param claim The PaymentMerkleClaim to be processed.
+     * Contains the root index, earner, payment leaves, and required proofs
+     * @dev only callable by the valid claimer, that is
+     * if claimerFor[claim.earner] is address(0) then only the earner can claim, otherwise only
+     * claimerFor[claim.earner] can claim the payments.
+     */
+    function processClaim(
+        PaymentMerkleClaim calldata claim
+    ) external onlyWhenNotPaused(PAUSED_CLAIM_PAYMENTS) nonReentrant {
+        DistributionRoot memory root = distributionRoots[claim.rootIndex];
+        _checkClaim(claim, root);
+        // If claimerFor earner is not set, claimer is by default the earner. Else set to claimerFor
+        address earner = claim.earnerLeaf.earner;
+        address claimer = claimerFor[earner];
+        if (claimer == address(0)) {
+            claimer = earner;
+        }
+        require(msg.sender == claimer, "PaymentCoordinator.processClaim: caller is not valid claimer");
+        for (uint256 i = 0; i < claim.tokenIndices.length; i++) {
+            TokenTreeMerkleLeaf calldata tokenLeaf = claim.tokenLeaves[i];
+
+            uint256 currCumulativeClaimed = cumulativeClaimed[earner][tokenLeaf.token];
+            require(
+                tokenLeaf.cumulativeEarnings >= currCumulativeClaimed,
+                "PaymentCoordinator.processClaim: cumulativeEarnings must be gte than cumulativeClaimed"
+            );
+
+            // Calculate amount to claim and update cumulativeClaimed
+            uint256 claimAmount = tokenLeaf.cumulativeEarnings - currCumulativeClaimed;
+            cumulativeClaimed[earner][tokenLeaf.token] = tokenLeaf.cumulativeEarnings;
+
+            tokenLeaf.token.safeTransfer(claimer, claimAmount);
```

Comment:
I think we can save gas by reducing the number of calls to `safeTransfer` since these all towards the same claimer.
The value can be aggregated to a local variable.

```
uint256 totalClaimAmount = 0;
for (uint256 i = 0; i < claim.tokenIndices.length; i++) {
            TokenTreeMerkleLeaf calldata tokenLeaf = claim.tokenLeaves[i];

            // Calculate amount to claim and update cumulativeClaimed
            uint256 claimAmount = tokenLeaf.cumulativeEarnings - currCumulativeClaimed;
            cumulativeClaimed[earner][tokenLeaf.token] = tokenLeaf.cumulativeEarnings;
            totalClaimAmount += claimAmount;
        }
        tokenLeaf.token.safeTransfer(claimer, totalClaimAmount);
```

Reply:
Plus, a simple batch function for the multiple earners with the same claimer would help too :)

```
function processClaim(PaymentMerkleClaim[] calldata claims)
```

Reply:
Interesting, so batch claiming for different earners, but assuming they all have the same claimer

Reply:
yes. @8sunyuan we would set the claimer for the +22k eigenpods to the same address and let it claim on behalf.

OR if the operator fee can be set to 100% I think we can play with it.